admin/services/service-manager/pupil-annulment/pupil-annulment.data.service.ts

```diff
@@ -0,0 +1,135 @@
+import sqlService, { TYPES } from '../../../services/data-access/sql.service'
+
+export class PupilAnnulmentDataService {
+  public static readonly annulmentCode = 'ANLLD'
+
+  static async setAnnulmentByUrlSlug (pupilUrlSlug: string, currentUserId: number): Promise<void> {
+    const params = [{
+      name: 'annuledAttendanceCode',
+      type: TYPES.Char(5),
+      value: PupilAnnulmentDataService.annulmentCode
+    }, {
+      name: 'pupilUrlSlug',
+      type: TYPES.UniqueIdentifier,
+      value: pupilUrlSlug
+    }, {
+      name: 'userId',
+      type: TYPES.Int,
+      value: currentUserId
+    }]
+    const sql = `
+    -- Pupil Annulment transaction
+
+    DECLARE @attendanceCode_id Int = (SELECT id from [mtc_admin].[attendanceCode] where code = @annuledAttendanceCode)
+    DECLARE @pupilId Int = (SELECT id from [mtc_admin].[pupil] where urlSlug = @pupilUrlSlug)
+
+    IF @attendanceCode_id IS NULL
+      THROW 51000, 'unknown attendanceCode.code', 1;
+
+    DECLARE @currentPupilCheckCompleteValue bit
+    DECLARE @currentPupilRestartAvailableValue bit
+
+    SELECT
+      @currentPupilCheckCompleteValue = checkComplete,
+      @currentPupilRestartAvailableValue = restartAvailable
+    FROM
+      [mtc_admin].[pupil]
+    WHERE
+      id = @pupilId
+
+    -- get id of any unconsumed restart
+    DECLARE @currentUnconsumedRestartId int
+
+    SELECT TOP 1
+      @currentUnconsumedRestartId = pr.id
+    FROM
+      [mtc_admin].[pupilRestart] pr
+    WHERE
+      pr.isDeleted = 0
+    AND
+      pr.pupil_id = @pupilId
+    AND
+      pr.check_id IS NULL
+    ORDER BY 1 DESC
```

Comment:
Please enlighten me: what does `ORDER BY 1 DESC` actually do?

Reply:
it returns the last entry, as its ordering by the column at ordinal 0, the primary key.